# Data Wrangling : Clinical trial data for a new oral insulin

Data wrangling is the process of assessing, structuring and cleaning raw data into a desired format for better decision making in less time. Data wrangling is increasingly ubiquitous at today’s top firms. Data has become more diverse and unstructured, demanding increased time spent culling, cleaning, and organizing data ahead of broader analysis. 


## Table of Contents
- [Introduction](#intro) <br>
<br>
- [Part I - Gathering data](#gather) <br>
<br>
- [Part II - Assessing data](#Assessing data)
    - [II-1 - Patients dataset](#Patients dataset)
    - [II-2 - Treatments dataset](#Treatments dataset)
    - [II-3 - Adverse_reactions dataset](#Adverse_reactions dataset)
    - [II-4 - Quality issues summary](#Quality issues summary)
    - [II-5 - Tidiness issues summary](#Tidiness issues summary) <br>
    <br>
- [Part III - Cleaning data](#Cleaning data)
    - [III-1 - Missing data](#Missing data)
    - [III-2 - Tidiness](#tidiness)
    - [III-3 - Quality](#quality)


<a id='intro'></a>
## Introduction

The increasing prevalence of diabetes in the 21st century is a problem, an epidemic even. Pre 1920s, diabetes was a feared disease, that most certainly led to death. Luckily, in the 1920s, a secretion in the pancreas that lowered blood sugar levels, soon to be called insulin, was discovered by soon to be Nobel Prize winner Frederick Banting.

People with severe diabetes, and only days left to live were saved, but the default method of administration back then was a needle, multiple times a day, and it still is now. This is scary for some people and uncomfortable and inconvenient for the vast majority. The future : Oral insulin. This is an active area of research, and has been for a long time. Historically though there's been a big roadblock, getting insulin through the stomachs thick lining.

In this project, we've got our hands on the phase two clinical trial data for a new innovative oral insulin called auralin. Auralin researchers believe their proprietary capsule will solve this stomach lining problem. Phase two trials test the efficacy and the dose response of a drug, plus identify common short term side effects, also known as adverse reactions. These typically involve several hundred patients.

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks


<a id='gather'></a>
## Part I - Gathering data

This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. Acquaint yourself with them through visual assessment below. <br>

To get started, let's import our libraries.

In [587]:
import pandas as pd

In [588]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

<a id='Assessing data'></a>
## Part II - Assessing data
<a id='Patients dataset'></a>
 ### II.1 Patients dataset


In [589]:
# Display the patients table first rows
patients.head(2)

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   

                                 contact  birthdate  weight  height   bmi  
0   951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1  PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [590]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


- Missing demographic informations ( adress, city , state, zip_code, country, contact)
- Erroneous datatype ( assigned_sex, state, zip_code, birthdate)

In [591]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

- zip code is a float instead of string 
- zip code sometimes is 4 digits instead of 5 ( minimum value 1002)
- height variable minimum value is 27 inches !!! given that the inclusion criteria of this clinical trial is 18 years old this doesn't seems like a plausible adult height.
- weight variable minimum value is 48.8 lbs !!! given that the inclusion criteria of this clinical trial is 18 years old this doesn't seems like a plausible adult weight.



In [592]:
# Let's check the the patient with the incorrect height of 27 inches
patients.query("height == 27")

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state  zip_code        country                          contact  birthdate  \
4    AL   36303.0  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  
4   192.3      27  26.1

Let's calculate the BMI if height is 27  and weight is 192.3

In [593]:
weight_lbs = 192.3
height_in = 27
BMI = 703 * weight_lbs / (height_in*height_in)
BMI

185.44156378600823

A BMI of 185 it's just can't be possible. Maybe it's a data entry error : 27 instead of 72

In [594]:
height_in = 72
BMI_check = 703 * weight_lbs / (height_in*height_in)
BMI_check

26.077719907407406

That make sens !

In [595]:
# the patient with the incorrect weight of 48 lbs
patients.query("weight == 48.8")

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

it seems like the weight is in kilograms instead of pounds Let's check this (

In [596]:
weight_lbs = 48.8*2.20462 # 2.20462 is the concersion factor between kilograms and pounds
height_in = 63
BMI = 703 * weight_lbs / (height_in*height_in)
BMI

19.05582654774502

In [597]:
patients.query("weight == 48.8").bmi

210    19.1
Name: bmi, dtype: float64

the calculated bmi match the bmi from the dataset

In [598]:
# Exploring state column
patients.state.value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
OH            14
Illinois      14
LA            13
MI            13
Florida       13
OK            13
NJ            12
VA            11
IL            10
WI            10
MS            10
MN             9
IN             9
FL             9
TN             9
AL             9
WA             8
KY             8
NC             8
MO             7
NV             6
KS             6
ID             6
IA             5
SC             5
CT             5
CO             4
RI             4
AR             4
Nebraska       4
ME             4
ND             4
AZ             4
DE             3
WV             3
SD             3
OR             3
MD             3
NE             2
VT             2
DC             2
MT             2
NM             1
NH             1
WY             1
AK             1
Name: state, dtype: int64

there are two representation name for some states : the full name and the abreviation. these states are
   -California : CA
   -New York : NY
   -Illinois : IL
   -Florida : FL
   -Nebraska  : NE            

In [599]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

- value count for surname 'Doe' is 6
- Multiple records for Jakobsen, Taylor, Gersten

In [600]:
patients.contact.sample(5)

68             414-646-4353HarunaAzuma@gustr.com
14             AsiaWozniak@rhyta.com918-712-3469
99     BerndSchneider@jourrapide.com914-830-3940
89          KwasiMensa@cuvox.de+1 (480) 967-6336
308        631-875-3023DaudBatukayev@teleworm.us
Name: contact, dtype: object

- tidiness issue : there are 2 variables in contact column : e-mail and phone number

<a id='Treatments dataset'></a>
 ### II.2 Treatments dataset

In [601]:
# Display the treatments table two first rows
treatments.head(2)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   

   hba1c_change  
0           NaN  
1          0.97


`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [602]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


- Erroneous datatype ( auralin, novodra)

In [603]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

-  a hba1c_change value about 0.4 is a success, however maximum value of hba1c_change is : 0.99 and 3rd quartile value is : 0.92 that suggest a massive skew and a big change wish is just implausible. <br>
Let's take an example and check it :

In [604]:
treatments.query('hba1c_change == 0.99').head(1)

given_name      surname auralin    novodra  hba1c_start  hba1c_end  \
32      laura  ehrlichmann       -  43u - 40u         7.95       7.46   

    hba1c_change  
32          0.99

In [605]:
# calculate hba1c_change
hba1c_start = 7.95
hba1c_end = 7.46
hba1c_change = hba1c_start - hba1c_end
hba1c_change

0.4900000000000002

- hba1c_change is 0.49 instead of 0.99 it is calculated wrong. this can be a data entry error. (4s mistaken as 9s)

In [606]:
# missing values in treatments table
treatments.isnull().sum()

given_name        0
surname           0
auralin           0
novodra           0
hba1c_start       0
hba1c_end         0
hba1c_change    109
dtype: int64

* Missing values in hba1c_change

In [607]:
treatments[['auralin', 'novodra']].head()

auralin    novodra
0  41u - 48u          -
1          -  40u - 45u
2          -  39u - 36u
3  33u - 36u          -
4          -  33u - 29u

- The letter 'u' is starting on ending doses for Auralin and Novodra 

In [608]:
patients[['given_name' , 'surname']].head()

given_name  surname
0        Zoe  Wellish
1     Pamela     Hill
2        Jae   Debord
3       Liêm     Phan
4        Tim  Neudorf

In [609]:
treatments[['given_name' , 'surname']].head()

given_name     surname
0   veronika    jindrová
1     elliot  richardson
2   yukitaka    takenaka
3       skye  gormanston
4     alissa      montez

- given name and surname in treatments table are all lowercase, but in the patients table they are not : this can be an issue if we want to join the two table using given_name and surname as a key

In [610]:
treatments.shape[0]

280

- missing records (280 instead of 350)


In [611]:
patients.contact.sample(5)

92            609-914-8473MichaelSmith@armyspy.com
100                  701-676-6301IsacBerg@cuvox.de
191    FanetteFournier@einrot.com+1 (312) 864-5737
118       337-685-4885AdibMutazzGhanem@fleckens.hu
445           MaximusHenzen@einrot.com408-792-9489
Name: contact, dtype: object

- there are different representation for telephone number

In [612]:
treatments.auralin.isnull().sum()

0

In [613]:
treatments.novodra.isnull().sum()

0

the two lines above suggest that there are zero null entries for both auralin and novodra columns. However there are some entries with dashes in both columns wish should be nulls

In [614]:
treatments.sample(5)

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
188      bouke      glaser          -  26u - 30u         7.72       7.34   
137      ivona      jakšić          -  41u - 41u         7.98       7.54   
1       elliot  richardson          -  40u - 45u         7.56       7.09   
114       sabr       amari          -  32u - 27u         7.94       7.57   
44       sauli  koivuniemi  43u - 47u          -         7.67       7.37   

     hba1c_change  
188           NaN  
137           NaN  
1            0.97  
114           NaN  
44           0.30

tidiness issues :  for auralin and novodra columns there are 3 variables:
        - treatment : auralin or novodra
        - starting dose
        - ending dose
        

<a id='Adverse_reactions dataset'></a>
 ### II.3 Adverse_reactions dataset

In [615]:
# Display the adverse_reactions table
adverse_reactions.head(2)

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

In [616]:
patients[['given_name' , 'surname']].head()

given_name  surname
0        Zoe  Wellish
1     Pamela     Hill
2        Jae   Debord
3       Liêm     Phan
4        Tim  Neudorf

In [617]:
adverse_reactions[['given_name' , 'surname']].head()

given_name     surname
0      berta  napolitani
1       lena        baer
2     joseph         day
3     flavia  fiorentino
4    manouck     wubbels

- given name and surname in adverse reactions table are all lowercase, but in the patients table they are not : this can be an issue if we want to join the two table using given_name and surname as a key

- adverse_reaction columns of adverse_reactions table belongs to the treatments table beacause an adverse reaction of a treatment is naturally connected to this treatment : we should join treatments table and adverse reactions table

In [618]:
# checking duplicated columns in three tables 
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

- we should remove given_name and surnme from the treatment table and keep patients_id as primary key to ensure joining tables

<a id='Quality issues summary'></a>
###  II.4 Quality issues summary <br>

#### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

#### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

#### `adverse_reactions` table
- Lowercase given names and surnames

<a id='Tidiness issues summary'></a>
### II.5 Tidiness issues summary<br>


Three rules to check tidiness <br>

__- Each variable forms a column__ <br>
__- Each observation forms a row__ <br>
__- Each type of observational unit forms a table__ <br> <br>

- contact column in patients table should be split into phone number and e-mail columns
- tree variables in two columns in treatments table (treatment, start_dose, end_dose)
- adverse_reaction columns of adverse_reactions table belongs to the treatments table
- given_name and surname columns are duplicated in patients and treatments table


<a id='Cleaning data'></a>
## Part III - Cleaning data <br>


Cleaning data is the third and final step in the data wrangling process. This is where the quality and tidiness issues identified in the assess step are remedied. It can be done manually in spreadsheet programs or text editors, but data cleaning is often best done using code and in three steps.

- __Define__ how to clean issue in words,
- __Code__ : convert these words to code,
- __Test__ to make sure if that code worked.



The very first thing to do before any cleaning occurs is to make a copy of each piece of data

In [619]:
# Making a copy of each dataset
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

<a id='Missing data'></a>
### III- 1 Missing Data

It's usually best to find missing data first if you can. In other words, completeness issues,  which is a part of data quality.
For our clinical trial dataset here, we identified three completeness issues.

Two in the treatments table: <br> 
- the missing HbA1c changes, and 
- the missing records having 280 instead of 350 <br>

Then one missing data issue in the patients table: 
- the missing demographic information (address, city, state, zip code, country, and contact)


We just can't address this third issue : the missing demographic information. We really have no way of getting this information from these people, because We can't contact them.

We can clean the other two issues though, the two in the treatments table: Missing HbA1c changes and the missing records (280 instead of 350)


### `treatments`: Missing records (280 instead of 350)


The missing `treatments` records are stored in a file named `treatments_cut.csv`. Let's read this file

In [620]:
# Import the cut treatments into a DataFrame
treatments_cut = pd.read_csv('treatments_cut.csv')

In [621]:
# Display the treatments_cut table first rows
treatments_cut.sample(5)

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
6        sara      miles  30u - 35u          -         7.53       7.12   
49    satsita  batukayev          -  42u - 42u         7.63       7.25   
60      maret   sultygov          -  26u - 23u         7.67       7.30   
47     regolo      nucci  51u - 59u          -         7.53       7.02   
17       emma     barlow  48u - 57u          -         7.85       7.39   

    hba1c_change  
6            NaN  
49          0.38  
60           NaN  
47           NaN  
17           NaN

In [622]:
# Number of rows and columns
treatments_cut.shape

(70, 7)

#### Define



- Concatenate treatments_clean and treatments_cut datasets using pandas.concat function ( [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) for pandas.concat function used in the solution)

#### Code

In [623]:
frames = [treatments_clean , treatments_cut]
treatments_clean = pd.concat(frames,ignore_index=True )

#### Test

In [624]:
treatments_clean.shape

(350, 7)

### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
*Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one **Define, Code, and Test** header occurs multiple times in this notebook.*

#### Define

Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

#### Code

In [625]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

#### Test

In [626]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
given_name      350 non-null object
surname         350 non-null object
auralin         350 non-null object
novodra         350 non-null object
hba1c_start     350 non-null float64
hba1c_end       350 non-null float64
hba1c_change    350 non-null float64
dtypes: float64(3), object(4)
memory usage: 19.2+ KB


In [627]:
treatments_clean.hba1c_change.describe()

count    350.000000
mean       0.396286
std        0.059283
min        0.200000
25%        0.350000
50%        0.390000
75%        0.440000
max        0.540000
Name: hba1c_change, dtype: float64

<a id='tidiness'></a>
### III- 2 Tidiness

After addressing missing data first, cleaning for tidiness is usually the next logical step.
For this oral insulin clinical trial dataset, we detected and documented four sections of the dataset that weren't tidy: 

- in the patients table, there are two variables in one column (contact column) : phone number and email 
- in the treatment's table, there are three variables in two columns (treatment, start dose and end dose in the auralin and novodra columns).
- the adverse_reactions table shouldn't exist ( the adverse reaction column should be part of the treatments table)
- the given_name and surname columns are duplicated in the treatments and in adverse_reactions table when it should only be in the patients table.

### Contact column in `patients` table contains two variables: phone number and email

#### Define

Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done. 

#### Code

We are going to use regular expressions with pandas' [`str.extract` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.extract.html). Here is an amazing [regex tutorial](https://regexone.com/). <br>


In [628]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

#### Test

In [629]:
# Confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [630]:
patients_clean.phone_number.sample(15)

333         260 623 1176
408    +1 (989) 390-0285
134    +1 (518) 212-7081
475         650-962-7179
362         757-624-1525
275    +1 (224) 305-6805
481         918 706 2776
390         979-639-7844
334         630-837-9473
415         909-861-4356
500         816-223-6007
470         386-234-5932
32          562-985-4582
399    +1 (360) 965-1558
239         228-378-1355
Name: phone_number, dtype: object

In [631]:
# Confirm that no emails start with an integer 
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

#### Define

Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

#### Code

In [632]:
# id_vars : Here we have set the variables (columns) that we want to leave unaffected. 
# Variables not included in this list will become rows in a new column (which has the name given by “var_name”)
# The values belonging to the original rows/columns are found in a new column with a name given by “value_name”
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [633]:
#we apply this mask to subset only data with available start and end dose 
treatments_clean = treatments_clean[treatments_clean.dose != "-"]

In [634]:
# split dose column
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str


In [635]:
# remove dose dose column
treatments_clean = treatments_clean.drop('dose', axis=1)

#### Test

In [636]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

### Adverse reaction should be part of the `treatments` table

##### Define

Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

#### Code

In [637]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')

#### Test

In [638]:
# Confirm the merge was executed correctly
treatments_clean.sample(10)

given_name    surname  hba1c_start  hba1c_end  hba1c_change  \
227              abel    yonatan         7.88       7.50          0.38   
139             alwin   svensson         7.78       7.34          0.44   
29       onyekachukwu     obinna         7.58       7.12          0.46   
224             kwasi      mensa         8.29       7.89          0.40   
117              arne  jørgensen         9.65       9.31          0.34   
55   zikoranaudodimma   chinedum         8.19       7.83          0.36   
108            taylah   mobourne         7.90       7.51          0.39   
72              lewis       webb         7.85       7.52          0.33   
163            regolo      nucci         7.53       7.02          0.51   
347             lixue      hsueh         9.21       8.80          0.41   

    treatment dose_start dose_end           adverse_reaction  
227   novodra        38u      39u                      cough  
139   auralin        36u      39u                        NaN  
29    auralin        37u      46u                        NaN  
224   novodra        40u      36u                        NaN  
117   auralin        32u      43u                        NaN  
55    auralin        41u      51u                        NaN  
108   auralin        41u      50u                        NaN  
72    auralin        39u      44u                        NaN  
163   auralin        51u      59u                        NaN  
347   novodra        22u      23u  injection site discomfort

### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
Adverse reactions table is no longer needed. 
- Isolate the patient ID and names in the `patients` table, then convert these names to lower case to join with `treatments`.<br>
- Drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

#### Code

In [639]:
#create new dataframe form patients table 
name_id = patients_clean[['patient_id' , 'given_name' , 'surname']]
# apply lowercase
name_id.given_name = name_id.given_name.str.lower()
name_id.surname = name_id.surname.str.lower()

In [640]:
# merge tables
treatments_clean = pd.merge(treatments_clean, name_id, on =['given_name' , 'surname'])

In [641]:
# drop given_name and surname columns
treatments_clean = treatments_clean.drop(['given_name' , 'surname'] , axis = 1)

#### Test

In [642]:
# Confirm the merge was executed correctly
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin        41u      48u   
1         7.97       7.62          0.35   auralin        33u      36u   
2         7.65       7.27          0.38   auralin        37u      42u   
3         7.89       7.55          0.34   auralin        31u      38u   
4         7.76       7.37          0.39   auralin        30u      36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

In [643]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

<a id='quality'></a>
### III- 3 Quality

Once the missing data and tidiness issues are cleaned, cleaning the remaining data quality issues, i.e.the ones outside of the completeness issues is all that remains.

There were 19 data quality issues in this dataset (16 outside of the missing data issues). Next we'll use a variety of common pandas functions to clean them all.

### Zip code is a float not a string and Zip code has four digits sometimes

#### Define

- Convert the zip code column's data type from a float to a string using (astype)
- Remove the .0 using string slicing and pad four digit zip codes witha leading 0

#### Code

In [644]:
# using str.pad to add padding (whitespaces or other characters) to every string element
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')

#### Test

In [645]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

### Tim Neudorf height is 27 in instead of 72 in

#### Define

- Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

#### Code

In [646]:
patients_clean['height'].replace(27, 72,inplace=True)


#### Test

In [647]:
# Should be empty
patients_clean.query('height == "27"')

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [648]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf'].height

4    72
Name: height, dtype: int64

### Full state names sometimes, abbreviations other times

#### Define

- Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska. <br>
This is a [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) for method used in solution.

#### Code

In [649]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

#### Test

In [650]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
MI    13
OK    13
LA    13
NJ    12
VA    11
MS    10
WI    10
MN     9
TN     9
AL     9
IN     9
NC     8
WA     8
KY     8
MO     7
NE     6
KS     6
NV     6
ID     6
IA     5
CT     5
SC     5
AZ     4
ND     4
ME     4
RI     4
CO     4
AR     4
DE     3
OR     3
MD     3
WV     3
SD     3
DC     2
MT     2
VT     2
NH     1
WY     1
NM     1
AK     1
Name: state, dtype: int64

### Dsvid Gustafsson

#### Define

Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

#### Code

In [651]:
patients_clean.given_name.replace('Dsvid','David', inplace = True)

#### Test

In [652]:
# Should be empty
patients_clean.query(' given_name == "Dsvid"')

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [653]:
patients_clean.query('surname == "Gustafsson"').given_name

8    David
Name: given_name, dtype: object

### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

#### Define
Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer. <br>

Hint methods used in solution: <br>
Method 1 :[documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) <br>
Method 2 :[documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html)<br>
Method 3 :[documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) 

#### Code

In [662]:
# Convert assigned sex and state to categorical data types
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')
#  Convert birthdate to datetime data type
patients_clean['birthdate'] = pd.to_datetime(patients_clean['birthdate'])
# Strip the letter 'u' in start dose and end dose and convert those columns to data type integer
treatments_clean['dose_start'] = treatments_clean['dose_start'].str.strip('u').astype(int)
treatments_clean['dose_end'] = treatments_clean['dose_end'].str.strip('u').astype(int)

#### Test

In [663]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [666]:
treatments_clean[['dose_start' , 'dose_end']].head()

dose_start  dose_end
0          41        48
1          33        36
2          37        42
3          31        38
4          30        36

In [667]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 342 entries, 0 to 341
Data columns (total 8 columns):
hba1c_start         342 non-null float64
hba1c_end           342 non-null float64
hba1c_change        342 non-null float64
treatment           342 non-null object
dose_start          342 non-null int32
dose_end            342 non-null int32
adverse_reaction    35 non-null object
patient_id          342 non-null int64
dtypes: float64(3), int32(2), int64(1), object(2)
memory usage: 21.4+ KB


### Multiple phone number formats

#### Define
- Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).
helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).*

#### Code

In [668]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

#### Test

In [669]:
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

### Default John Doe data

##### Define
- Remove the non-recoverable John Doe records from the `patients` table.

#### Code

In [670]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

#### Test

In [673]:
# Should be no Doe records
patients_clean[patients_clean.surname == 'Doe']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

### Multiple records for Jakobsen, Gersten, Taylor

#### Define

- Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

#### Code

In [674]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

#### Test

In [675]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [676]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [677]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

### kgs instead of lbs for Zaitseva weight

#### Define

Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

#### Code

In [678]:
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

#### Test

In [680]:
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.min()

102.09999999999999